src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,443 @@
+from typing import Any, List, Tuple
+
+import numpy as np
+from numpy.typing import NDArray
+
+from frequenz.sdk.timeseries._moving_window import MovingWindow
+
+log = logging.getLogger(__name__)
+
+
+class PeriodicFeatureExtractor:
+    """
+    A tool that extracts a timeseries feature out of historical data.
+
+    This class is creating a profile from periodically occurring
+    windows in a buffer of historical data.
+
+    The profile is created out of all windows that are fully
+    contained in the underlying buffer with the same
+    start and end time modulo a fixed period.
+
+    Note that the oldest window or the window that is currently overwritten in
+    the MovingWindow is not considered in the profile.
+
+    Note further that when constructing a PeriodicFeatureExtractor object
+    the `MovingWindow` size has to be a integer multiple of
+    the period.
+
+    Example:
+        ```python
+            feature_extractor = PeriodicFeatureExtractor(
+            moving_window = moving_window, # 28 days moving window
+            distance=timedelta(days=7),
+        )
+
+        now = datetime.now(tzinfo=timezone.utc)
+
+        # create a daily weighted average for the next 24h
+        avg_24h = feature_extractor.avg(
+            now, now + timedelta(hours=24),
+            weights=[0.1, 0.2, 0.3, 0.4]
+        )
+
+        # create a daily average for Thursday March 23 2023
+        th_avg_24h = feature_extractor(datetime(2023, 03, 23), datetime(2023, 03, 24))
+        ```
+    """
+
+    def __init__(
+        self,
+        moving_window: MovingWindow,
+        period: timedelta,
+    ) -> None:
+        """
+        Initialize a PeriodicFeatureExtractor object.
+
+        Args:
+            moving_window: The MovingWindow that is used for the average calculation.
+            period: The distance between two succeeding intervals.
+
+        Raises:
+            ValueError: If the MovingWindow size is not a integer multiple of the period.
+        """
+        self._moving_window = moving_window
+        self._buffer = moving_window._buffer
+
+        # internally we are working with indexes of samples
+        self._sampling_period = self._moving_window.sampling_period.total_seconds()
+
+        # distance between two succeeding intervals in samples
+        self._distance = int(period.total_seconds() / self._sampling_period)
+
+        log.debug("len(moving_window): %i", len(self._moving_window))
+        log.debug("distance: %i", self._distance)
+
+        if not len(self._moving_window) % self._distance == 0:
+            raise ValueError(
+                "The MovingWindow size is not a integer multiple of the distance."
+            )
+
+        if self._distance - period.total_seconds() / self._sampling_period != 0:
+            log.warning(
+                "The distance is not a multiple of the sampling period. "
+                "This might result in unexpected behaviour."
+            )
```

Comment:
+1, I actually thought the same when reading this but assumed there was. And if there is, it would be good to state it explicitly in a comment, so people looking at the code in the future don't ask themselves the same question.